packages/@webex/plugin-cc/src/services/core/WebSocket/WebSocketManager.ts

```diff
@@ -0,0 +1,176 @@
+    await this.connect().catch(() => {
+      LoggerProxy.logger.error(`[WebSocketStatus] | Error in connecting Websocket`);
+    });
+  }
+
+  close(shouldReconnect: boolean, reason = 'Unknown') {
+    if (!this.isSocketClosed && this.shouldReconnect) {
+      this.shouldReconnect = shouldReconnect;
+      this.websocket.close();
+      this.keepaliveWorker.postMessage({type: 'terminate'});
+      LoggerProxy.logger.error(
+        `[WebSocketStatus] | event=webSocketClose | WebSocket connection closed manually REASON: ${reason}`
+      );
+    }
+  }
+
+  private async register(connectionConfig: SubscribeRequest) {
+    try {
+      const subscribeResponse: SubscribeResponse = await this.webex.request({
+        service: WCC_API_GATEWAY,
+        resource: SUBSCRIBE_API,
+        method: HTTP_METHODS.POST,
+        body: connectionConfig,
+      });
+      this.url = subscribeResponse.body.webSocketUrl;
+    } catch (e) {
+      LoggerProxy.logger.error(
+        `Register API Failed, Request to RoutingNotifs websocket registration API failed ${e}`
+      );
+    }
+  }
+
+  private async connect() {
+    if (!this.url) {
+      return undefined;
+    }
+    LoggerProxy.logger.log(
+      `[WebSocketStatus] | event=webSocketConnecting | Connecting to WebSocket: ${this.url}`
+    );
+    this.websocket = new WebSocket(this.url);
+
+    return new Promise((resolve, reject) => {
+      this.websocket.onopen = () => {
+        this.isSocketClosed = false;
+        this.shouldReconnect = true;
+
+        this.websocket.send(JSON.stringify({keepalive: 'true'}));
+        this.keepaliveWorker.onmessage = (keepAliveEvent: {data: any}) => {
+          if (keepAliveEvent?.data?.type === 'keepalive') {
+            this.websocket.send(JSON.stringify({keepalive: 'true'}));
+          }
+
+          if (keepAliveEvent?.data?.type === 'closeSocket' && this.isConnectionLost) {
+            this.forceCloseWebSocketOnTimeout = true;
+            this.close(true, 'WebSocket did not auto close within 16 secs');
+            LoggerProxy.logger.error(
+              '[webSocketTimeout] | event=webSocketTimeout | WebSocket connection closed forcefully'
+            );
+          }
+        };
+
+        this.keepaliveWorker.postMessage({
+          type: 'start',
+          intervalDuration: KEEPALIVE_WORKER_INTERVAL, // Keepalive interval
+          isSocketClosed: this.isSocketClosed,
+          closeSocketTimeout: CLOSE_SOCKET_TIMEOUT, // Close socket timeout
+        });
+      };
+
+      this.websocket.onerror = (event: any) => {
+        LoggerProxy.logger.error(
+          `[WebSocketStatus] | event=socketConnectionFailed | WebSocket connection failed ${event}`
+        );
+        reject();
+      };
+
+      this.websocket.onclose = async (event: any) => {
+        this.webSocketOnCloseHandler(event);
+      };
+
+      this.websocket.onmessage = (e: MessageEvent) => {
+        this.dispatchEvent(new CustomEvent('message', {detail: e.data}));
```

Comment:
As per discussion with Ravi, I have switched over to using events here as opposed to singals. I felt this was a more straight-forward solution. Now, I am not using custom event handlers here as it was working with events, however I would like someone to point out any edge cases or issues that might arise (most things are working including reconnect logic).